client/src/app/users/user.service.spec.ts

```diff
@@ -59,135 +60,249 @@ describe('User service: ', () => {
+      // name: Chris and Jamie.
+      expect(filteredUsers.length).toBe(2);
+      // Every returned user's name should contain an 'i'.
+      filteredUsers.forEach(user => {
+        expect(user.name.indexOf(userName)).toBeGreaterThanOrEqual(0);
+      });
+    });
 
-  it('filterUsers() filters by company', () => {
-    expect(testUsers.length).toBe(3);
-    const userCompany = 'UMM';
-    expect(userService.filterUsers(testUsers, { company: userCompany }).length).toBe(1);
-  });
+    it('filters by company', () => {
+      const userCompany = 'UMM';
+      const filteredUsers = userService.filterUsers(testUsers, { company: userCompany });
+      // There should be just one user that has UMM as their company.
+      expect(filteredUsers.length).toBe(1);
+      // Every returned user's company should contain 'UMM'.
+      filteredUsers.forEach(user => {
+        expect(user.company.indexOf(userCompany)).toBeGreaterThanOrEqual(0);
+      });
+    });
 
-  it('filterUsers() filters by name and company', () => {
-    expect(testUsers.length).toBe(3);
-    const userCompany = 'UMM';
-    const userName = 'chris';
-    expect(userService.filterUsers(testUsers, { name: userName, company: userCompany }).length).toBe(1);
+    it('filters by name and company', () => {
+      // There's only one user (Chris) whose name
+      // contains an 'i' and whose company contains
+      // an 'M'. There are two whose name contains
+      // an 'i' and two whose company contains an
+      // an 'M', so this should test combined filtering.
+      const userName = 'i';
+      const userCompany = 'M';
+      const filters = { name: userName, company: userCompany };
+      const filteredUsers = userService.filterUsers(testUsers, filters);
+      // There should be just one user with these properties.
+      expect(filteredUsers.length).toBe(1);
+      // Every returned user should have _both_ these properties.
+      filteredUsers.forEach(user => {
+        expect(user.name.indexOf(userName)).toBeGreaterThanOrEqual(0);
+        expect(user.company.indexOf(userCompany)).toBeGreaterThanOrEqual(0);
+      });
+    });
   });
 
-  it('addUser() posts to api/users', () => {
-
-    userService.addUser(testUsers[1]).subscribe(
-      id => expect(id).toBe('testid')
-    );
-
-    const req = httpTestingController.expectOne(userService.userUrl);
-
-    expect(req.request.method).toEqual('POST');
-    expect(req.request.body).toEqual(testUsers[1]);
-
-    req.flush({id: 'testid'});
+  describe('Adding a user using `addUser()`', () => {
+    it('talks to the right endpoint and is called once', waitForAsync(() => {
+      // Mock the `httpClient.addUser()` method, so that instead of making an HTTP request,
+      // it just returns our test data.
+      const USER_ID = 'pat_id';
+      const mockedMethod = spyOn(httpClient, 'post').and.returnValue(of(USER_ID));
+
+      // paying attention to what is returned (undefined) didn't work well here,
```

Comment:
Are you saying that the post request is expected to return undefined? That shouldn't be the case. I don't know whether this is really blocking or not, but it seems weird to get an undefined anywhere here.

Reply:
Yeah, I don't think you shouldn't have gotten `undefined` back here. If you did, then perhaps the mocking wasn't complete? I don't immediately see the problem, though, but it does seem like something's wrong.

Reply:
I've now created an issue to explore this further.

Reply:
Thinking back to when I first wrote this comment (maybe for lab 3?), I think it was not something that was undefined in the test or mock, but in the console it was undefined. I think I'll need to have some help figuring out what we actually want to do here.